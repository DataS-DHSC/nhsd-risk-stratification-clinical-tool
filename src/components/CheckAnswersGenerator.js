import { Button } from 'nhsuk-react-components';
import PropTypes from 'prop-types';
import AnswerListRow from './AnswerListRow';
import HiddenStateHolder from './HiddenStateHolder';
import { addNamespace } from '../utils/namespacingTools';

const CheckAnswersGenerator = ({ prevData, forms, scoreUrl, formBaseUrl }) => {
  const hasInvalidValues = Object.entries(forms)
    .flatMap(([formId, form]) => [[formId, form.sections[0]]])
    .map(([formId, section]) => [formId, Object.entries(section.fields)])
    .flatMap(([formId, sectionArray]) =>
      sectionArray.map((field) => [formId, field])
    )
    .filter(([, [, field]]) => Boolean(field.fieldValidator))
    .map(([, [fieldId, field]]) =>
      field.fieldValidator(prevData[addNamespace(fieldId)])
    )
    .some((message) => message.length > 0);

  return (
    <div className="nhsuk-grid-row">
      <div className="nhsuk-grid-column-two-thirds">
        <h1 id="check-answers-heading">Check your answers</h1>
        <p id="check-information-is-correct-paragraph">
          This is the information that will be used to generate the
          patient&apos;s result. Please check that each answer you have given is
          correct.
        </p>

        <form id="check-answers-form" action={scoreUrl} method="POST">
          <HiddenStateHolder stateFields={prevData} />
          {Object.entries(forms).map(([formKey, form]) => (
            <AnswerListRow
              key={`answer-check-${formKey}`}
              pageTemplate={form}
              data={prevData}
              formBaseUrl={formBaseUrl}
            />
          ))}
          <h2 id="generate-risk-score-heading">
            Now generate the patient&apos;s result
          </h2>
          <p id="submit-disclaimer-paragraph">
            By submitting this notification you are confirming that, to the best
            of your knowledge, the details you are providing are correct.
          </p>
          <Button
            type="submit"
            disabled={hasInvalidValues}
            className="nhsuk-u-margin-bottom-8"
            id="continue-button"
          >
            Accept and generate result
          </Button>
        </form>
      </div>
    </div>
  );
};

CheckAnswersGenerator.propTypes = {
  prevData: PropTypes.objectOf(PropTypes.string),
  forms: PropTypes.shape().isRequired,
  scoreUrl: PropTypes.string.isRequired,
  formBaseUrl: PropTypes.string.isRequired,
};

CheckAnswersGenerator.defaultProps = {
  prevData: {},
};

export default CheckAnswersGenerator;
