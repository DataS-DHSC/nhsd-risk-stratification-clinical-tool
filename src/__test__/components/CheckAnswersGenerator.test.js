import React from 'react';
import { render } from '@testing-library/react';
import CheckAnswersGenerator from '../../components/CheckAnswersGenerator';

const TEST_FORMS = {
  'test-form': {
    label: 'form1',
    sections: [
      {
        label: 'section1',
        fields: {
          field1: {
            fieldValidator: () => '',
          },
          field2: {
            fieldValidator: null,
          },
        },
      },
    ],
  },
};

const setup = ({ forms }) => {
  const props = {
    forms,
    prevData: {
      'answers.field1': 'test1',
      'answers.field2': 'test2',
    },
    fieldDefaults: {
      'test-form': {
        field1: '',
        field2: '',
      },
    },
    scoreUrl: '/score',
    formBaseUrl: '/ras/algoName',
  };

  return render(<CheckAnswersGenerator {...props} />);
};

describe('CheckAnswersGenerator', () => {
  it('should match snapshot when all validations pass', () => {
    const container = setup({ forms: TEST_FORMS });
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when some validations fails', () => {
    const forms = { ...TEST_FORMS };
    forms['test-form'].sections[0].fields.field1.fieldValidator = () =>
      'validation message';
    const container = setup({ forms });

    expect(container.asFragment()).toMatchSnapshot();
  });
});
