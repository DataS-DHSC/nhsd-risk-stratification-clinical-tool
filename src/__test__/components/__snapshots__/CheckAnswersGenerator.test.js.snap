// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckAnswersGenerator should match snapshot when all validations pass 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        id="check-answers-heading"
      >
        Check your answers
      </h1>
      <p
        id="check-information-is-correct-paragraph"
      >
        This is the information that will be used to generate the patient's result. Please check that each answer you have given is correct.
      </p>
      <form
        action="/score"
        id="check-answers-form"
        method="POST"
      >
        <input
          name="answers.field1"
          type="hidden"
          value="test1"
        />
        <input
          name="answers.field2"
          type="hidden"
          value="test2"
        />
        <h2
          id="answer-list-form1-heading"
        >
          form1
        </h2>
        <h3
          id="answer-list-section1-heading"
        >
          section1
        </h3>
        <dl
          class="nhsuk-summary-list"
        >
          <div
            class="nhsuk-summary-list__row"
            id="summary-row-field1"
          >
            <dt
              class="nhsuk-summary-list__key"
            />
            <dd
              class="nhsuk-summary-list__value"
            >
              test1
            </dd>
            <dd
              class="nhsuk-summary-list__actions"
            >
              <button
                aria-disabled="false"
                aria-label="Change undefined from test1"
                class="nhsuk-button change-button"
                formaction="/ras/algoName/undefined"
                id="change-button-field1"
                type="submit"
              >
                Change
              </button>
            </dd>
          </div>
          <div
            class="nhsuk-summary-list__row"
            id="summary-row-field2"
          >
            <dt
              class="nhsuk-summary-list__key"
            />
            <dd
              class="nhsuk-summary-list__value"
            >
              test2
            </dd>
            <dd
              class="nhsuk-summary-list__actions"
            >
              <button
                aria-disabled="false"
                aria-label="Change undefined from test2"
                class="nhsuk-button change-button"
                formaction="/ras/algoName/undefined"
                id="change-button-field2"
                type="submit"
              >
                Change
              </button>
            </dd>
          </div>
        </dl>
        <h2
          id="generate-risk-score-heading"
        >
          Now generate the patient's result
        </h2>
        <p
          id="submit-disclaimer-paragraph"
        >
          By submitting this notification you are confirming that, to the best of your knowledge, the details you are providing are correct.
        </p>
        <button
          aria-disabled="false"
          class="nhsuk-button nhsuk-u-margin-bottom-8"
          id="continue-button"
          type="submit"
        >
          Accept and generate result
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CheckAnswersGenerator should match snapshot when some validations fails 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        id="check-answers-heading"
      >
        Check your answers
      </h1>
      <p
        id="check-information-is-correct-paragraph"
      >
        This is the information that will be used to generate the patient's result. Please check that each answer you have given is correct.
      </p>
      <form
        action="/score"
        id="check-answers-form"
        method="POST"
      >
        <input
          name="answers.field1"
          type="hidden"
          value="test1"
        />
        <input
          name="answers.field2"
          type="hidden"
          value="test2"
        />
        <h2
          id="answer-list-form1-heading"
        >
          form1
        </h2>
        <h3
          id="answer-list-section1-heading"
        >
          section1
        </h3>
        <dl
          class="nhsuk-summary-list"
        >
          <div
            class="nhsuk-summary-list__row"
            id="summary-row-field1"
          >
            <dt
              class="nhsuk-summary-list__key"
            />
            <dd
              class="nhsuk-summary-list__value nhsuk-validation-error"
            >
              test1 (validation message)
            </dd>
            <dd
              class="nhsuk-summary-list__actions"
            >
              <button
                aria-disabled="false"
                aria-label="Change undefined from test1 (validation message)"
                class="nhsuk-button change-button"
                formaction="/ras/algoName/undefined"
                id="change-button-field1"
                type="submit"
              >
                Change
              </button>
            </dd>
          </div>
          <div
            class="nhsuk-summary-list__row"
            id="summary-row-field2"
          >
            <dt
              class="nhsuk-summary-list__key"
            />
            <dd
              class="nhsuk-summary-list__value"
            >
              test2
            </dd>
            <dd
              class="nhsuk-summary-list__actions"
            >
              <button
                aria-disabled="false"
                aria-label="Change undefined from test2"
                class="nhsuk-button change-button"
                formaction="/ras/algoName/undefined"
                id="change-button-field2"
                type="submit"
              >
                Change
              </button>
            </dd>
          </div>
        </dl>
        <h2
          id="generate-risk-score-heading"
        >
          Now generate the patient's result
        </h2>
        <p
          id="submit-disclaimer-paragraph"
        >
          By submitting this notification you are confirming that, to the best of your knowledge, the details you are providing are correct.
        </p>
        <button
          aria-disabled="true"
          class="nhsuk-button nhsuk-button--disabled nhsuk-u-margin-bottom-8"
          disabled=""
          id="continue-button"
          type="submit"
        >
          Accept and generate result
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
